proxy.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+)
+
+type proxyResponse struct {
+	resp *http.Response
+	err  error
+}
+
+var errProxyTimeout = errors.New("all peers timed out")
+
+// proxyRequest proxies the request, trying each peer that should have the key
+// in turn. The total logical attempt will be capped at the configured proxy
+// timeout, but individual peers will be tried using the following algorithm:
+//   - Each interval of 'proxy_stage_timeout', starting immediately, a request
+//     is kicked off to one random not-yet-tried peer. All requests after
+//     the first run concurrently.
+//   - If a request finishes successfully to any peer, the result is returned
+//     immediately and the others are canceled, as long as the result was not
+//     an error.
+//   - If a request finishes, but resulted in an error, another is kicked off
+//     immediately to another random not-yet-tried peer.
+//   - This process continues until either all peers are being tried, in which
+//     case the code just waits for one to finish. If the total 'proxy_timeout'
+//     is hit at any point, the method returns immediately with an error and
+//     cancels any running requests.
+func (vs *version) proxyRequest(r *http.Request, peers []string) (*http.Response, error) {
+	responses := make(chan proxyResponse, len(peers))
+	cancel := make(chan struct{})
+
+	peerIndex := 0
+	outstanding := 0
+	totalTimeout := time.NewTimer(vs.sequins.config.Sharding.ProxyTimeout.Duration)
+	for {
+		stageTimeout := time.NewTimer(vs.sequins.config.Sharding.ProxyStageTimeout.Duration)
+		if peerIndex < len(peers) {
+			url := fmt.Sprintf("http://%s%s?proxy=%s", peers[peerIndex], r.URL.Path, vs.name)
+			go vs.proxyAttempt(url, responses, cancel)
+			peerIndex += 1
+			outstanding += 1
+		} else if outstanding == 0 {
+			return nil, errNoAvailablePeers
+		}
+
+		select {
+		case res := <-responses:
+			if res.err != nil {
+				log.Printf("Error proxying request to peer: %s", res.err)
+				outstanding -= 1
+				continue
+			} else {
+				close(cancel)
+				return res.resp, nil
+			}
+		case <-totalTimeout.C:
+			close(cancel)
+			return nil, errProxyTimeout
+		case <-stageTimeout.C:
+		}
+	}
+
+	return nil, errNoAvailablePeers
+}
```

Comment:
It would be cool to keep counters of

* when we kick off a backup request
* when a backup request wins
* when all of them fail and we return `errNoAvailablePeers`